### PR TITLE
Adding jamovi to flathub

### PR DESCRIPTION
**Pull request by jonathon-love**
hi,

here is our flatpak for jamovi (note that jamovi is stylised with a lowercase 'j')

i've done what i think is right, but i appreciate there may be places where i've misunderstood things, and am keen to correct things where i've erred.

a few things i wasn't sure about:

1. should i be referring to paths with `$FLATPAK_DEST/...` or just `app/...`. i wasn't sure which should be used where.

2. the `rename-desktop-file` at al. options. i wasn't quite sure how these worked, or how/if i'm supposed to use them (i couldn't quite figure out from looking at examples either). do i copy the `jamovi.desktop` file to `/app/usr/share/menu/jamovi.desktop`, and then the `rename-desktop-file` renames it to `org.jamovi.jamovi.desktop`? copying it across with the correct filename to begin with seemed simpler, so i figured i don't need to use it?

3. appstream. `appstream-util validate-relax` outputs:

```
jamovi.appdata.xml: FAILED:
• markup-invalid        : <id> does not have correct extension for kind
• attribute-invalid     : <screenshot> width too large [https://www.jamovi.org/misc/screenshot-linux-anova.png]
• attribute-invalid     : <screenshot> height too large [https://www.jamovi.org/misc/screenshot-linux-anova.png]
• attribute-invalid     : <screenshot> width too large [https://www.jamovi.org/misc/screenshot-linux-descriptives.png]
• attribute-invalid     : <screenshot> height too large [https://www.jamovi.org/misc/screenshot-linux-descriptives.png]
Validation of files failed
```

   googling it, it looks like i can ignore the `markup-invalid` error.

   wrt screenshot size, they're normal sized screenshots taken on a modern hi-res screen, but the appstream website says that we should not scale down our screenshots. so i felt caught between conflicting directives. i also couldn't find anywhere anything on the maximum size.

   i did wonder if perhaps these issues were caused by an older version of appstream-util, but the build instructions for (what i think is the appstream-util repo) didn't work for my distro.

4. desktop integration
   i couldn't figure out how a flatpak integrates itself into the 'start menu' of the distro. normally the .desktop file does this, but i'm not sure how this works with flatpaks. also, the .desktop file i provide has the old reference to `/usr/bin/jamovi`, but i can't see that being correct any more now it's a flatpak, but i couldn't find other people patching their .desktop files in other flatpak examples either. so i'm unsure how this works. any guidance would be appreciated.

5. xml2-config. a *lot* of R packages make use of `xml2-config`, and so i created my own `xml2-config` which wraps `pkg-config`. this seemed simpler than patching *a lot* of R packages. hopefully this is an acceptable solution.

6. python. i included python in the package, because that was [the only way i could get to build on my machine](https://github.com/flatpak/flatpak-builder-tools/issues/16#issuecomment-368372987). but appreciate this is included in the runtime, and can be removed.

anyhow, sorry i wasn't able to figure these things out. super keen for feedback.

with thanks

jonathon

### CODE REVIEW

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

**Comment by nedrichards** [comment]
Wow, there's a lot to review here! Thanks for the notes ;-) I'll try and work through it. One initial thought is that this build will fail immediately on arm and aarch64 since you're not building electron for them. However, for Electron 1.8.2 there is arm and aarch64 binaries upstream. If you have other binary dependencies that aren't available on that architecture you need to add a flathub.json file to prevent the build system from trying those architectures.

**Comment by jonathon-love** [comment]
> Wow, there's a lot to review here!

i know! i'm hoping there's an award for having the most ridiculous amount of dependencies!

> One initial thought is that this build will fail immediately on arm and aarch64 since you're not building electron for them. However, for Electron 1.8.2 there is arm and aarch64 binaries upstream

ah yup. i wasn't sure what architectures were built on flathub (or what the names of the architectures were) can you point me to some documentation for this? the docs [here](http://docs.flatpak.org/en/latest/flatpak-command-reference.html) explain the `arches` type options, but not the possible values (i think i figured out `i386` by looking at examples). is it `arm` and `arm64` respectively?

> If you have other binary dependencies that aren't available on that architecture you need to add a flathub.json file to prevent the build system from trying those architectures.

righto. again, is there some docs on `flathub.json`? or wait, i just found this:

https://github.com/flathub/flathub/wiki/App-Maintenance

it explains the architectures too, so i think i'm set.

with thanks

**Comment by jonathon-love** [comment]
OK, i've added arm and arm64. it seems unlikely me that *all* of our dependencies will build for these arches, but i guess we'll see what happens.

with thanks

**Comment by nedrichards** [comment]
Flatpak calls them arm and aarch64, in fact we just added some quick mapping to some of the npm helper tools the other day: https://github.com/flatpak/flatpak-builder-tools/commit/b7b3be8424d810342a67cd5c98c42bf04b03caf6

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

**Comment by nedrichards** [comment]
Linking to the failed build https://flathub.org/builds/#/builders/1/builds/2466

**Comment by jonathon-love** [comment]
hi,

it turns out the CRAN urls (for R packages) aren't stable. fortunately microsoft provide snapshot mirrors of CRAN repos, so i've transitioned to using that instead (you might want to make a note of this, because other projects using R packages will face similar issues).

cheers

ps. can i instruct the build bot? i'll try

**Comment by jonathon-love** [comment]
bot, build org.jamovi.jamovi

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

**Comment by nedrichards** [comment]
Only repo admins can instruct the build bot. Once the app is added you'll be able to use it yourself to build and test PRs, which is pretty handy.

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

**Comment by nedrichards** [comment]
https://flathub.org/builds/#/builders/2/builds/3

**Comment by barthalion** [comment]
bot, build org.jamovi.jamovi

**Comment by jonathon-love** [comment]
hey, wrt to the x86_64 error, it doesn't look like something i'm responsible for (or have i broken some rule which has interfered with the flathub build system)?

```
stripping /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp/files/lib/python3.5/site-packages/scipy/sparse/csgraph/_shortest_path.cpython-35m-x86_64-linux-gnu.so to /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp/files/lib/debug/lib/python3.5/site-packages/scipy/sparse/csgraph/_shortest_path.cpython-35m-x86_64-linux-gnu.so.debug
FB: Running: eu-strip --remove-comment --reloc-debug-sections -f /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp/files/lib/debug/lib/python3.5/site-packages/scipy/sparse/csgraph/_shortest_path.cpython-35m-x86_64-linux-gnu.so.debug -F /app/lib/debug/lib/python3.5/site-packages/scipy/sparse/csgraph/_shortest_path.cpython-35m-x86_64-linux-gnu.so.debug /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp/files/lib/python3.5/site-packages/scipy/sparse/csgraph/_shortest_path.cpython-35m-x86_64-linux-gnu.so
]2;flatpak-builder: Cleanup python3-scipyFB: unmounting rofiles-fuse /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp
Committing stage build-python3-scipy to cache
]2;flatpak-builder: Updating python3-scipyFB: starting: rofiles-fuse /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/builddir /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp
Error: Can't spawn rofiles-fuseFailed to execute child process "rofiles-fuse" (No such file or directory)
FB: unmounting rofiles-fuse /mnt/builder/worker/build-x86_64-3/org.jamovi.jamovi/.flatpak-builder/rofiles/rofiles-43tHpp
program finished with exit code 1
elapsedTime=3465.694211
```

**Comment by jonathon-love** [comment]
i'm now trying to reproduce the i386 build error, do i install the electron base app for i386 with:

```
flatpak install flathub io.atom.electron.BaseApp --arch i386
```

because i get a:

```
error: Error searching remote flathub: Can't find ref org.atom.electron.BaseApp/i386
```

any tips?

with thanks

**Comment by jonathon-love** [comment]
i got it:

```
flatpak install flathub io.atom.electron.BaseApp/i386/stable
```

**Comment by TingPing** [comment]
bot, build org.jamovi.jamovi

**Comment by TingPing** [comment]
> should i be referring to paths with $FLATPAK_DEST/... or just app/.... i wasn't sure which should be used where.

Both are valid, `/app` is more readable IMO. `$FLATPAK_DEST` is mostly used when the file is used by both a runtime and an application.

> i couldn't figure out how a flatpak integrates itself into the 'start menu' of the distro. normally the .desktop file does this, but i'm not sure how this works with flatpaks. also, the .desktop file i provide has the old reference to /usr/bin/jamovi, but i can't see that being correct any more now it's a flatpak, but i couldn't find other people patching their .desktop files in other flatpak examples either. so i'm unsure how this works. any guidance would be appreciated.

Flatpak will *export* any `.desktop` file you install that has the prefix of your app id. So `/app/share/applications/org.jamovi.jamovi.desktop` will be exposed to the host. `flatpak` rewrites the contents of all these so they run in the sandbox.

> xml2-config. a lot of R packages make use of xml2-config, and so i created my own xml2-config which wraps pkg-config. this seemed simpler than patching a lot of R packages. hopefully this is an acceptable solution.

Sounds reasonable.

> python. i included python in the package, because that was the only way i could get to build on my machine. but appreciate this is included in the runtime, and can be removed.

Yea I'm really not a fan of that

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

**Comment by jonathon-love** [comment]
ooh, sorry. i've still got one more commit to go. this won't build.

**Comment by jonathon-love** [comment]
OK, this should be good to build. sorry for stuffing you around.

**Comment by TingPing** [comment]
bot, build org.jamovi.jamovi

**Comment by jonathon-love** [comment]
OK, this should be good to go.

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

**Comment by jonathon-love** [comment]
no idea what was wrong with the app-data, or why the most recent commit fixes it, but it does.

this is good to go.

**Comment by nedrichards** [comment]
bot, build org.jamovi.jamovi

File: org.jamovi.jamovi.json [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+                },
+                {
+                    "type": "git",
+                    "url": "https://github.com/jamovi/jamovi-compiler.git",
+                    "tag": "v0.8.3",
+                    "commit": "435545f8841b1374495a3176b16ed0a11a203a47",
+                    "dest": "jamovi-compiler"
+                },
+                {
+                    "type": "script",
```
Comment (barthalion) [inline org.jamovi.jamovi.json]: Sounds l ike something we should put into runtime…

**Comment by barthalion** [comment]
Repository has been created: https://github.com/flathub/org.jamovi.jamovi

Make sure to accept the invitation to collaboration. Thanks!

**Comment by jonathon-love** [comment]
thanks guys!

so the `flathub/org.jamovi.jamovi` repo now exists. what happens from here? (and now we've got the build working, can i squash all those commits into one?)

with thanks

**Comment by barthalion** [comment]
The `master` branch is protected from force pushes so no (and it's actually valuable to see possible fixes when packaging something else). You are able to push new commits now, and you can trigger test builds for pull requests by putting `bot, build` into comment (without app id like we were doing here).

Jamovi will be visible on Flathub applications listing when someone updates appdata, although new "shop" is being working on that will be more dynamic in that regard.

**Comment by TingPing** [comment]
It is visible here: https://beta.flathub.org/apps/details/org.jamovi.jamovi

There is an issue getting the icon though that will need to be fixed in the package probably.

**Comment by jonathon-love** [comment]
> There is an issue getting the icon though that will need to be fixed in the package probably.

is there a way i can find out more about this? double checking everything &ndash; everything seems to be in order.

with thanks

**Comment by TingPing** [comment]
>   <id>org.jamovi.jamovi</id>

Hmm, there was an old `flatpak` bug for that style of id, but I think the builders updated.

>                 "sed -E -e 's|Icon=.*|Icon=/app/share/icons/hicolor/scalable/apps/org.jamovi.jamovi.svg|g' platform/jamovi.desktop > /app/share/applications/org.jamovi.jamovi.desktop",

Remove that, just install the file and set the `"rename-desktop-file": "jamovi.desktop"` property on the package.

Ideally rename it upstream =)

>                 "cp platform/app-icon.svg /app/share/icons/hicolor/scalable/apps/org.jamovi.jamovi.svg",

Same comment but use `"rename-icon": "jamovi"`. This will fix the icon name in the desktop file also.

While you are at it make sized icons:
```json
         "for size in 64 128; do
              rsvg-convert -w $size -h $size -f png -o $size.png /app/share/icons/hicolor/scalable/apps/jamovi.svg
              install -Dm644 $size.png /app/share/icons/hicolor/${size}x${size}/apps/jamovi.png
            done"
```

>                 "cp platform/menu /app/share/menu/org.jamovi.jamovi",

Unrelated but What is that? I imagine it does nothing?

**Comment by jonathon-love** [comment]
> Remove that, just install the file and set the "rename-desktop-file": "jamovi.desktop" property on the package.

righto. the 'start menu' entry has the following:

```
/usr/bin/flatpak run --branch=stable --arch=x86_64 --command=/usr/bin/jamovi --file-forwarding org.jamovi.jamovi @@ %f @@
```

it looks like it's taking the `--command` from the `.desktop` file. do i need to patch the .desktop file accordingly? (i would have expected it to use the `command` specified in the manifest)

**Comment by TingPing** [comment]
> it looks like it's taking the --command from the .desktop file. do i need to patch the .desktop file accordingly? (i would have expected it to use the command specified in the manifest)

Yes, fix the desktop file upstream to not hardcode the wrong path.

**Pull request closed without merging**